### DISCUSSION TITLE
Define a documentation standard.

### DISCUSSION THREAD

**Discussion by SebastianJL** [typst/typst#2705] [General]
Hello everyone.

After using Typst for about a month now and increasingly writing my own code I came to the conclusion that it might be sensible to define how elements like functions and variables should be documented.

Advantages of doing this now:
- Keeps the community efforts streamlined.
- LSPs could support this standard and render the information.
- Later either Typst or a third party tool could support automatic documentation generation for typst packages.

While researching about this topic I found [https://github.com/Mc-Zen/tidy](https://github.com/Mc-Zen/tidy) that already allows generation of documentation from typst source code.
It looks a lot like the official documentation on [https://typst.app/docs](https://typst.app/docs):

> ![image](https://github.com/typst/typst/assets/11479035/691bdd78-adcd-450d-803c-1f660a668b9f)

I'm not saying that the recommended standard should look exactly like this. I'm saying that it makes sense to think about a standard now and document it on the website.
Maybe @MC-Zen would like to chime in.

**Comment by Dherse** [comment]
This has been discussed before and afaik it's planned, but this requires some more work on the way several parts of typst work and I think it shouldn't be expected before we have user declared elements and types (which are coming).

**Comment by Mc-Zen** [comment]
Hi, yes we discussed this a bit some time ago and there were some disagreements on how the comments should look like. I agree with @Dherse that waiting for types and co. makes sense since this will profoundly change how we look at Typst code. Also, (optional) in-code type annotations were proposed by @laurmaedje once. This _might_ make it unnecessary to annotate types in the docstring (which is nice I think). 

That being said, it would be good to think about this rather soon than late to avoid too much confusion and different standards, in my opinion. 

True, at some point, we will also need a (probably CLI) docs generator that is more flexible because in pure Typst we cannot generate any files, except for the document output obviously (although I may say that tidy is about to get some more nice updates that allow for example code execution and light-weight docstring tests). For now, tidy aims to fill the gap; but of course another important thing will be generation of web docs and this can naturally not be achieved with Typst and a single output file. 

However the standard may look in the end, I vote for `///` strings rather than the Java-style
```
/**
 *
 */
```
😂 (which someone also proposed). Mainly because it is much harder to parse and not entirely unambigous (the star may also belong to the content in some cases). Simple `//` comments can easily lead to (more serious) ambiguity as well (is it a mere comment put in a line in front of some declaration or is it a docstring?). But it in the end, this is just a detail.

**Comment by SebastianJL** [comment]
Thanks for your answers. Is there a particular reason we couldn't start with a smaller standard for functions and variables now and include elements and types as they come?

Type annotations would be wonderful. I would even prefer if they where checked against at compile time as this gives nicer errors at the function boundary instead of inside the function. It's easier for the user if a function fails at the call site when the wrong kind of argument is used instead of inside the template code.

For example, I don't really like how type annotations work in python. You can sense that they are an afterthought. For example it's often hard to express the types you intend to accept in python functions.

Ok, but I'm drifting off. Sorry for that.
